File: scripts/tfog/render.py
"""Turn plan.json into the text a reviewer reads on the PR.

Mirrors internal/plan/summary.go. Two rules from there are reproduced exactly, because both are
about the number people actually read:

  * A replace is an action *pair* — ["delete","create"], or ["create","delete"] under
    create_before_destroy — not its own action string. Counting it as one create plus one delete
    understates destruction in the headline.
  * `addresses` is the default detail level, and it is a security default rather than a
    terseness one. A plan file embeds a snapshot of prior state, so every secret in state is in
    the plan, and Terraform's `sensitive` marking covers what the schema declares — not a token
    somebody pasted into an unmarked metadata attribute.
"""

from __future__ import annotations

import json

# internal/plan/summary.go caps at 200 for the same reason: a GitHub comment body is limited
# (65536 bytes) and oversize payloads are rejected rather than truncated, so bound the list
# before formatting and say how many were dropped.
MAX_RENDERED_ADDRESSES = 200
MAX_FULL_DIFF_CHARS = 30000

SYMBOL = {
    "create": "+",
    "update": "~",
    "delete": "-",
    "replace": "-/+",
    "read": "<=",
}


def counts(plan_json: bytes | dict) -> dict:
    doc = plan_json if isinstance(plan_json, dict) else json.loads(plan_json)
    tally = {"create": 0, "update": 0, "delete": 0, "replace": 0, "read": 0}
    for rc in doc.get("resource_changes") or []:
        kind = classify(rc)
        if kind:
            tally[kind] += 1
    return tally


def classify(resource_change: dict) -> str | None:
    actions = list((resource_change.get("change") or {}).get("actions") or [])
    if not actions or actions == ["no-op"]:
        return None
    if sorted(actions) == ["create", "delete"]:
        return "replace"
    if actions == ["create"]:
        return "create"
    if actions == ["update"]:
        return "update"
    if actions == ["delete"]:
        return "delete"
    if actions == ["read"]:
        return "read"
    return None


def has_changes(tally: dict) -> bool:
    return tally["create"] + tally["update"] + tally["delete"] + tally["replace"] > 0


def headline(tally: dict) -> str:
    parts = [
        f"{tally['create']} to add",
        f"{tally['update']} to change",
        f"{tally['delete']} to destroy",
    ]
    if tally["replace"]:
        parts.append(f"{tally['replace']} to replace")
    return ", ".join(parts)


def address_lines(plan_json: bytes | dict, limit: int = MAX_RENDERED_ADDRESSES) -> tuple[list[str], int]:
    doc = plan_json if isinstance(plan_json, dict) else json.loads(plan_json)
    lines = []
    for rc in doc.get("resource_changes") or []:
        kind = classify(rc)
        if not kind or kind == "read":
            continue
        lines.append(f"{SYMBOL[kind]:>3} {rc.get('address', '?')}")
    lines.sort(key=lambda s: s.split(None, 1)[-1])
    omitted = max(0, len(lines) - limit)
    return lines[:limit], omitted


# ---------------------------------------------------------------------------
# Comment bodies
# ---------------------------------------------------------------------------


def plan_body(
    *,
    workspace,
    meta: dict,
    plan_json: bytes,
    plan_text: str,
    plan_path,
    tally: dict,
    marker: str,
) -> str:
    changed = has_changes(tally)
    lines, omitted = address_lines(plan_json)

    out = [marker, f"### `terraform plan` — **{workspace.name}**", ""]
    verdict = headline(tally) if changed else "**No changes.** Infrastructure matches the configuration."
    out.append(f"`{workspace.dir}` · {verdict}")
    out.append("")

    if lines:
        out.append("```diff")
        out.extend(lines)
        if omitted:
            out.append(f"… and {omitted} more resource change(s) not shown")
        out.append("```")
        out.append("")

    if workspace.summary_detail == "full" and changed:
        diff = plan_text
        note = ""
        if len(diff) > MAX_FULL_DIFF_CHARS:
            note = f"\n… truncated; {len(diff) - MAX_FULL_DIFF_CHARS} more characters in `plan.txt`"
            diff = diff[:MAX_FULL_DIFF_CHARS]
        out.append("<details><summary>Full diff</summary>")
        out.append("")
        out.append("```terraform")
        out.append(diff + note)
        out.append("```")
        out.append("")
        out.append("</details>")
        out.append("")

    out.append(
        f"plan `{meta['base_sha'][:8]}…{meta['head_sha'][:8]}` · terraform {meta['terraform_version']}"
        f" · {meta['duration_seconds']}s · planned by @{meta['planned_by']}"
    )
    out.append("")
    out.append(
        f"<sub>tree `{meta['planned_tree_sha'][:12]}` · plan `{meta['plan_sha256'][:12]}` · config from "
        f"`{meta['config_ref']}` @ `{meta['config_ref_sha'][:8]}` · artifacts `{plan_path}`</sub>")
    out.append("")
    if workspace.summary_detail != "full" and changed:
        out.append(
            "<sub>Only addresses and actions are shown: a plan file embeds a snapshot of prior "
            "state. The full diff is in `plan.txt` in the store above.</sub>"
        )
        out.append("")
    if not workspace.apply.enabled:
        out.append("<sub>This workspace is **plan-only** — merging will not apply it.</sub>")
        out.append("")
    out.append(_meta_block(meta))
    return "\n".join(out)


def apply_body(*, workspace, meta: dict, tally: dict, ok: bool, detail: str, log_path, marker: str) -> str:
    out = [marker, f"### `terraform apply` — **{workspace.name}**", ""]
    if ok:
        out.append(f"Applied the reviewed plan. {headline(tally)}.")
    else:
        out.append("**Apply failed.**")
    out.append("")
    if detail:
        out.append("```")
        out.append(detail.strip())
        out.append("```")
        out.append("")
    out.append(
        f"plan `{meta['base_sha'][:8]}…{meta['head_sha'][:8]}` · merge commit "
        f"`{(meta.get('merge_commit_sha') or '')[:8]}` · tree `{meta['planned_tree_sha'][:12]}` verified"
    )
    out.append("")
    out.append(f"<sub>terraform {meta['terraform_version']} · log `{log_path}`</sub>")
    if not ok:
        out.append("")
        out.append(
            "<sub>An apply that fails part-way has already changed some infrastructure. Read the "
            "log before re-running anything — the next plan shows what remains.</sub>"
        )
    return "\n".join(out)


def blocked_body(*, marker: str, title: str, detail: str) -> str:
    return "\n".join([marker, f"### {title}", "", detail, ""])


def _meta_block(meta: dict) -> str:
    """meta.json in an HTML comment: invisible in the timeline, readable by tfog-apply.

    This is the portable half of the key. tfog-apply locates the plan in the local store on its
    own, and cross-checks these values against it — so a forged comment can only make an apply
    *refuse*, never make it apply something else. That asymmetry is the point, and it is why
    the comment can be treated as an index without being treated as an authority.
    """
    return "<!-- tfog-meta\n" + json.dumps(meta, indent=2, sort_keys=True) + "\n-->"


def parse_meta_block(body: str) -> dict | None:
    start = body.find("<!-- tfog-meta")
    if start < 0:
        return None
    end = body.find("-->", start)
    if end < 0:
        return None
    try:
        return json.loads(body[start + len("<!-- tfog-meta") : end])
    except json.JSONDecodeError:
        return None

File: internal/prcomment/render.go
// Package prcomment renders the PR comments the local commands post.
//
// Separate from internal/plan, which renders check-run output for the services. The content is
// the same; the container is not, and the two have different size limits and different Markdown
// affordances. Keeping them apart means neither has to be written to suit the other.
package prcomment

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/sampleserve/terraform-on-github/internal/config"
	"github.com/sampleserve/terraform-on-github/internal/plan"
	"github.com/sampleserve/terraform-on-github/internal/store"
)

// metaOpen and metaClose delimit the machine-readable block.
const (
	metaOpen  = "<!-- tfog-meta"
	metaClose = "-->"
)

// PlanInput is everything a plan comment renders from.
type PlanInput struct {
	Workspace config.Workspace
	Meta      store.Meta
	Digest    string
	ShowJSON  []byte
	ShowText  string
	StorePath string
	Marker    string
	PlannedBy string
}

// Plan renders the sticky plan comment.
func Plan(in PlanInput) (string, error) {
	counts, err := plan.Counts(in.ShowJSON)
	if err != nil {
		return "", err
	}
	changed := counts.Any()

	var b strings.Builder
	fmt.Fprintf(&b, "%s\n### `terraform plan` — **%s**\n\n", in.Marker, in.Workspace.Name)

	verdict := plan.Headline(counts)
	if !changed {
		verdict = "**No changes.** Infrastructure matches the configuration."
	}
	fmt.Fprintf(&b, "`%s` · %s\n\n", in.Workspace.Dir, verdict)

	lines, omitted, err := plan.AddressLines(in.ShowJSON, plan.MaxRenderedAddresses)
	if err != nil {
		return "", err
	}
	if len(lines) > 0 {
		b.WriteString("```diff\n")
		for _, l := range lines {
			fmt.Fprintf(&b, "%s\n", l)
		}
		if omitted > 0 {
			fmt.Fprintf(&b, "… and %d more resource change(s) not shown\n", omitted)
		}
		b.WriteString("```\n\n")
	}

	if diff := plan.Redact(in.ShowText, in.Workspace.SummaryDetail); diff != "" && changed {
		fmt.Fprintf(&b, "<details><summary>Full diff</summary>\n\n```terraform\n%s\n```\n\n</details>\n\n", diff)
	}

	fmt.Fprintf(&b, "plan `%s…%s` · terraform %s · %s · planned by @%s\n\n",
		short(in.Meta.BaseSHA, 8), short(in.Meta.HeadSHA, 8), in.Meta.TerraformVersion,
		plan.HumanDuration(in.Meta.Duration), in.PlannedBy)

	fmt.Fprintf(&b, "<sub>tree `%s` · plan `%s` · config from `%s` @ `%s` · artifacts `%s`</sub>\n\n",
		short(in.Meta.PlannedTreeSHA, 12), short(in.Meta.PlanSHA256, 12),
		in.Meta.ConfigRef, short(in.Meta.ConfigRefSHA, 8), in.StorePath)

	if in.Workspace.SummaryDetail != config.SummaryFull && changed {
		b.WriteString("<sub>Only addresses and actions are shown: a plan file embeds a snapshot of prior " +
			"state. The full diff is in `plan.txt` in the store above.</sub>\n\n")
	}
	if !in.Workspace.ApplyEnabled() {
		b.WriteString("<sub>This workspace is **plan-only** — merging will not apply it.</sub>\n\n")
	}

	block, err := metaBlock(in.Meta, in.Digest)
	if err != nil {
		return "", err
	}
	b.WriteString(block)
	return b.String(), nil
}

// ApplyInput is everything an apply comment renders from.
type ApplyInput struct {
	Workspace config.Workspace
	Meta      store.Meta
	Counts    store.ResourceCounts
	OK        bool
	Detail    string
	LogPath   string
	Marker    string
}

// Apply renders the outcome comment. Posted fresh rather than edited in place, so the timeline
// reads as one story: plan, then what happened.
func Apply(in ApplyInput) string {
	var b strings.Builder
	fmt.Fprintf(&b, "%s\n### `terraform apply` — **%s**\n\n", in.Marker, in.Workspace.Name)

	if in.OK {
		fmt.Fprintf(&b, "Applied the reviewed plan. %s.\n\n", plan.Headline(in.Counts))
	} else {
		b.WriteString("**Apply failed.**\n\n")
	}
	if d := strings.TrimSpace(in.Detail); d != "" {
		fmt.Fprintf(&b, "```\n%s\n```\n\n", d)
	}
	fmt.Fprintf(&b, "plan `%s…%s` · merge commit `%s` · tree `%s` verified\n\n",
		short(in.Meta.BaseSHA, 8), short(in.Meta.HeadSHA, 8),
		short(in.Meta.MergeCommitSHA, 8), short(in.Meta.PlannedTreeSHA, 12))
	fmt.Fprintf(&b, "<sub>terraform %s · log `%s`</sub>\n", in.Meta.TerraformVersion, in.LogPath)

	if !in.OK {
		b.WriteString("\n<sub>An apply that fails part-way has already changed some infrastructure. " +
			"Read the log before re-running anything — the next plan shows what remains.</sub>\n")
	}
	return b.String()
}

// Notice renders a standalone advisory or blocking comment.
func Notice(marker, title, detail string) string {
	return fmt.Sprintf("%s\n### %s\n\n%s\n", marker, title, detail)
}

// ---------------------------------------------------------------------------
// The machine-readable block
// ---------------------------------------------------------------------------

// metaBlock embeds meta.json in an HTML comment: invisible in the timeline, readable by
// cmd/tfog-apply.
//
// This is the portable half of the key, and it is an index rather than an authority. tfog-apply
// locates the plan in the local store itself and only cross-checks these values, so a forged
// comment can make an apply *refuse* — never make it apply something else. That asymmetry is what
// lets the comment be trusted for lookup without being trusted for authorization.
func metaBlock(m store.Meta, digest string) (string, error) {
	payload := struct {
		store.Meta
		Digest string `json:"digest"`
	}{Meta: m, Digest: digest}
	body, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s\n%s\n%s", metaOpen, body, metaClose), nil
}

// Published is what a plan comment claims about the plan it describes.
type Published struct {
	store.Meta
	Digest string `json:"digest"`
}

// ParseMeta extracts the block from a comment body. Returns nil when there is none.
func ParseMeta(body string) *Published {
	start := strings.Index(body, metaOpen)
	if start < 0 {
		return nil
	}
	rest := body[start+len(metaOpen):]
	end := strings.Index(rest, metaClose)
	if end < 0 {
		return nil
	}
	var p Published
	if err := json.Unmarshal([]byte(rest[:end]), &p); err != nil {
		return nil
	}
	return &p
}

func short(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}
